src/rAF/index.ts

```diff
@@ -0,0 +1,49 @@
+import { Entity } from "../forces/Force.gen";
+
+interface RAFState {
+  lastFrame: DOMHighResTimeStamp;
+  animationFrameId: number | null;
+  listener: (
+    timestamp: DOMHighResTimeStamp,
+    lastFrame: DOMHighResTimeStamp,
+    mover: Entity
+  ) => Entity;
+  mover: Entity;
+}
+
+export const rAF = () => {
+  const state: RAFState = {
+    lastFrame: performance.now(),
+    animationFrameId: null,
+    listener: ((() => {}) as unknown) as RAFState["listener"],
+    mover: {
+      acceleration: [0, 0],
+      velocity: [0, 0],
+      position: [0, 0]
+    }
+  };
+
+  const start = (listener: RAFState["listener"]) => {
+    state.listener = listener;
+    draw(state.lastFrame);
+  };
+
+  const draw = (timestamp: DOMHighResTimeStamp) => {
+    state.mover = state.listener(timestamp, state.lastFrame, state.mover);
+    console.log(state.mover);
+    state.lastFrame = timestamp;
+    state.animationFrameId = requestAnimationFrame(timestamp => {
+      draw(timestamp);
+    });
+  };
```

Comment:
Won't the exiting condition be computed in the `state.listener` most of the time? probably some precision threshold for each physics step?

If so shouldn't it have some way of communicating whether or not to schedule the next frame?

Or do you have something else in mind with how you'll consume this that bypasses this problem?

Reply:
That's a really good point. I like the API you suggested in Slack, where `start` returns the `stop` function such that the `listener` has a way to cancel things. I suppose I haven't quite thought enough about the stopping aspect of things. Like, the gravitational force is theoretically going on for forever. I think `react-spring` has some `isOvershooting` flag to know when the spring has reach its end, so I could have something similar to determine a condition like "the mover has reached (or nearly reached) the attractor". In either case I think you're right that the `stop` will have to move, this just felt like a fine place to put it for the time being.